Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,310 @@
+
+## References
+
+* [N. Bourbaki, *General Topology*][bourbaki1966]
+
+## Tags
+
+Hausdorff, group action, proper action
+-/
+
+open Filter Topology Set Prod
+
+/-- Proper group action in the sense of Bourbaki:
+the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+@[mk_iff]
+class ProperVAdd (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [AddGroup G]
+    [AddAction G X] : Prop where
+  /-- Proper group action in the sense of Bourbaki:
+  the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+  isProperMap_vadd_pair' : IsProperMap (fun gx ↦ ⟨gx.1 +ᵥ gx.2, gx.2⟩ : G × X → X × X)
+
+/-- Proper group action in the sense of Bourbaki:
+the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+@[to_additive existing, mk_iff]
+class ProperSMul (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [Group G]
+    [MulAction G X] : Prop where
+  /-- Proper group action in the sense of Bourbaki:
+  the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+  isProperMap_smul_pair' : IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X)
+
+attribute [to_additive existing] properSMul_iff
+
+/-- By definition, if G acts properly on X then
+the map `G × X → X × X` is a proper map. -/
+@[to_additive "By definition, if G acts properly on X then
+the map `G × X → X × X` is a proper map."]
+lemma isProperMap_smul_pair (G X : Type*) [Group G] [MulAction G X]
+    [TopologicalSpace G] [TopologicalSpace X] [ProperSMul G X] :
+    IsProperMap (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) :=
+  ProperSMul.isProperMap_smul_pair'
+
+variable {G X Y Z : Type*} [Group G] [MulAction G X] [MulAction G Y]
+variable [TopologicalSpace G] [TopologicalSpace X] [TopologicalSpace Y]
+variable [TopologicalSpace Z]
+
+/-- If a group acts properly then in particularl it acts continuously. -/
+@[to_additive "If a group acts properly then in particularl it acts continuously."]
+instance continuousSmul_of_properSMul [ProperSMul G X] : ContinuousSMul G X where
+  continuous_smul := (isProperMap_smul_pair G X).continuous.fst
+
+/-- A group `G` acts properly on a topological space `X` if and only if for all ultrafilters
+`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `g • x₂ = x₁` and `𝒰.fst` converges to `g`. -/
+@[to_additive "A group acts `G` properly on a topological space `X` if and only if
+for all ultrafilters `𝒰` on `X`, if `𝒰` converges to `(x₁, x₂)`
+along the map `(g, x) ↦ (g • x, x)`, then there exists `g : G` such that `g • x₂ = x₁`
+and `𝒰.fst` converges to `g`."]
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto : ProperSMul G X ↔ ContinuousSMul G X ∧
+    (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, g • x₂ = x₁ ∧ Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · refine' fun h ↦ ⟨inferInstance, fun 𝒰 x₁ x₂ h' ↦ _⟩
```

Comment:
I see you've replaced a lot of `refine` by `refine'`, but I'm pretty sure we would like instead to get rid of `refine'`, which is a leftover from the port. The syntax of `refine` may look a bit ugly because of all the `?_`, but these actually have a nice semantic distinction: `_` means "Lean should find this alone" while `?_` means "I will prove this later". 

Additionally, note that you can use `?name` instead to create a named goal, which allows you to prove the goals in the order you want, but arguably this is a bit of a gadget.

Reply:
In fact I hesitated between the two because I thought `?_` avoided confusion with placeholders which were to be found by Lean but on the other hand felt a bit weird. But if we want to get rid of `refine'` then I replace it.